### PR TITLE
Support for skipping tests and assertions.

### LINKED ISSUE #105: Skip option for ut test case
Add a method for skipping a test case based on some condition.
```
ut test case("foo") << Skip If(Expr(...))
```

### PR DESCRIPTION
**Pull request by emccorkle**
Closes #105.

## Description
Adds support for skipping tests and individual assertions.

## Checklist

- [x] **I am adding a new matcher.**
  - [x] I have included documentation with at least the prototype and description in a comment above the matcher factory.
  - [x] I have included the prototype, description, and at least one example in the matcher factory definition for use in the Scripting Index.
  - [ ] I have added tests in `Tests/UnitTests/Matchers` (included in `Tests/UnitTests/SkipTest.jsl` instead)
  - [x] I have added matcher to `Source/Matchers-Index.txt`.
  - [x] I have added a note to `CHANGELOG.md`.

- [x] **I am adding new or changing current functionality.**
  - [x] I have added or updated the tests for the new or changed functionality in `Tests/UnitTests`.
  - [x] I have added note(s) to `CHANGELOG.md` as necessary.

## Contributing

- [x] I have read and agree to the [Contributor Agreement](https://github.com/sassoftware/jsl-hamcrest/blob/master/ContributorAgreement.txt)

Signed-off-by: Evan McCorkle <Evan.McCorkle@jmp.com>

### CODE REVIEW

File: Source/Addin/Reporters.jsl [hunk 1]
```diff
@@ -166,7 +173,10 @@ ut format node entry = Function({data}, {short desc, label details, str},
...
 			1, "<b>EXPRESSION FAILURE</b>  ^label details^",
 			2, "<b>FAILURE</b>  ^Xml Encode(Char(Name Expr(data[3])))^ ^Xml Encode(Char(short desc))^ ^label details^\!N    <i>Actual:</i> ^Xml Encode(Char(data[5]))^\!N    <i>Expected:</i> ^Xml Encode(Char(data[4]))^",
-			3, "<b>FAILURE</b>  ^Xml Encode(Char(Name Expr(data[3])))^ ^Xml Encode(Char(short desc))^ ^label details^\!N    <i>Actual:</i> threw &quot;^Xml Encode(Char(data[5]))^&quot;\!N    <i>Expected:</i> ^Xml Encode(Char(data[4]))^"
+			3, "<b>FAILURE</b>  ^Xml Encode(Char(Name Expr(data[3])))^ ^Xml Encode(Char(short desc))^ ^label details^\!N    <i>Actual:</i> threw &quot;^Xml Encode(Char(data[5]))^&quot;\!N    <i>Expected:</i> ^Xml Encode(Char(data[4]))^",
+		   -1, If(Is Empty(data[3]) & Length(data[4]) == 0,
+					"<b>SKIPPED</b>  Entire test ^label details^\!N    <i>Skipped</i>: ^Xml Encode(Char(data[5]))^",
+					"<b>SKIPPED</b>  ^Xml Encode(Char(Name Expr(data[3])))^ ^Xml Encode(Char(short desc))^ ^label details^\!N    <i>Skipped</i>: ^Xml Encode(Char(data[5]))^\!N    <i>Expected</i>: ^Xml Encode(Char(data[4]))^");
```
Comment (JustinChilton) [inline Source/Addin/Reporters.jsl]: Alignment looks a bit off here

File: Source/Addin/Reporters.jsl [hunk 2]
```diff
@@ -177,7 +187,10 @@ ut format node entry = Function({data}, {short desc, label details, str},
 			0, "SUCCESS  ^Name Expr(data[3])^ ^short desc^ ^label details^\!N    Expected: ^data[4]^",
 			1, "EXPRESSION FAILURE  ^label details^",
 			2, "FAILURE  ^Name Expr(data[3])^ ^short desc^ ^label details^\!N    Actual: ^data[5]^\!N    Expected: ^data[4]^",
-			3, "FAILURE  ^Name Expr(data[3])^ ^short desc^ ^label details^\!N    Actual: threw \!"^data[5]^\!"\!N    Expected: ^data[4]^"
+			3, "FAILURE  ^Name Expr(data[3])^ ^short desc^ ^label details^\!N    Actual: threw \!"^data[5]^\!"\!N    Expected: ^data[4]^",
+		   -1, If(Is Empty(data[3]) & Length(data[4]) == 0,
+					"SKIPPED  Entire test ^label details^\!N    Skipped: ^data[5]^",
+					"SKIPPED  ^Name Expr(data[3])^ ^short desc^ ^label details^\!N    Skipped: ^data[5]^\!N    Expected: ^data[4]^");
```
Comment (JustinChilton) [inline Source/Addin/Reporters.jsl]: Spacing looks off here too.

File: Source/Core/TestCase.jsl [hunk 3]
```diff
@@ -17,10 +17,15 @@ Define Class( "UtTestCaseFixture",
 	// Variable: name
 	// Name of the test case
 	name = "";
-	// Variable:setup expr
+	// Variable: setup expr
 	setup expr = Empty();
 	// Variable: teardown expr 
 	teardown expr = Empty();
+	// Variable: skip flag
+	skip expr = Expr(0);
```
Comment (JustinChilton) [inline Source/Core/TestCase.jsl]: This is just `0` since `Expr` doesn't do anything here. Would `Empty()` be better? Then when `<<Get Skip If` is called, you get `Empty()` instead of 0.

File: Source/Core/TestCase.jsl [hunk 4]
```diff
@@ -74,6 +79,35 @@ Define Class( "UtTestCaseFixture",
...
+		Arguments: 
+			skip expr - determine if all cases should be skipped (eval delayed until ut test)
+			skip description - human-readable description of the reason for skip (used if skipped)
+			
+		Returns:
+			Reference to this object
+	*/
+	Skip If = Method({skip expr, skip description=Empty()},
+		this:skip expr = Name Expr( skip expr );
+		this:skip description = If(Is Empty(skip description), Char(Name Expr(skip expr)), skip description);
```
Comment (JustinChilton) [inline Source/Core/TestCase.jsl:102]: Unfortunately, `IsEmpty()` returns true for an empty string. If someone passes in `""`, should we infer the description then? Otherwise, you can set the default value for `skip description` to numeric missing and check for `IsMissing()` instead.

File: Source/Core/TestCase.jsl [hunk 5]
```diff
@@ -219,6 +264,11 @@ Define Class("UtSuccessFilteringReporter",
 	add success = Method( {label, test expr, description, payload=Empty()},
 		.
 	);
+	add skip = Method( {label, test expr, description, skip description, payload=Empty()},
+		// Counted as a success for now.
+		// This is used for the log benchmarking which has an independent skipping mechanism.
```
Comment (JustinChilton) [inline Source/Core/TestCase.jsl]: Update `This` to `This Reporter`. It took me a second that it was about the class and not the method.

File: Source/Core/TestCase.jsl [hunk 6]
```diff
@@ -468,8 +531,56 @@ ut define documented function(
 			);
 		);
 		
-		// Get Do Log Bench() < 0 means don't capture
-		If( _utTestNS:log benchmark << Get Do Log Bench() >= 0,
+		// evaluate the skip if expression
+		_utTestNS:skip? =
+			Try(
```
Comment (JustinChilton) [inline Source/Core/TestCase.jsl]: Optional: This `Try(` could be moved up a line and section un-indented a tab.

File: Source/Matchers/Skip.jsl [hunk 7]
```diff
@@ -30,42 +30,79 @@ Define Class(
...
-		ut skip( UtMatcher matcher = ut anything() )
+		ut skip( UtMatcher matcher = ut anything(), why = "" )
 		---------------
 		Used to mark an expression you want to skip. Instead a success
-		is reported.
+		is reported. Usually, you want to give a reason why you are
+		skipping.
 		
-		Factory function for <UtSkipMatcher>.
+		In addiition, this is mostly useful as an outer matcher. Most
```
Comment (JustinChilton) [inline Source/Matchers/Skip.jsl]: addition*

**Review by JustinChilton — CHANGES_REQUESTED** [review]
Initial review.

File: Source/Addin/Reporters.jsl (hunk 2 above)
Comment (emccorkle) [reply Source/Addin/Reporters.jsl]: image

File: Source/Addin/Reporters.jsl (hunk 1 above)
Comment (emccorkle) [reply Source/Addin/Reporters.jsl]: image

File: Source/Core/TestCase.jsl (hunk 3 above)
Comment (emccorkle) [reply Source/Core/TestCase.jsl]: I can change it to just `0` or maybe `ut expr literal(0)`. We want it to be false so that the default value works well for the implicit test case created when you do `ut test("string", ...)`. I just used `Expr(0)` since I wanted to indicate that it was supposed to be an expression and that it is false by default.

File: Source/Core/TestCase.jsl (hunk 4 above)
Comment (emccorkle) [reply Source/Core/TestCase.jsl:102]: Good idea, I'll do this here and for `ut skip`.

Comment (emccorkle) [reply Source/Core/TestCase.jsl:102]: You had me convinced but `IsEmpty("")` returns `0`. Still need to update `ut skip` though.

Comment (JustinChilton) [reply Source/Core/TestCase.jsl:102]: My bad, it is `Is Missing("")` that returns `1`. Still like the new implementation better.

File: Source/Core/TestCase.jsl (hunk 3 above)
Comment (JustinChilton) [reply Source/Core/TestCase.jsl]: I like it as just `0`. Making it `ut expr literal(0)` would have literally put `ut expr literal(0)` as the expression, due to the evaluation behavior of Namespaces and Classes.

File: Source/Addin/Reporters.jsl (hunk 1 above)
Comment (JustinChilton) [reply Source/Addin/Reporters.jsl]: Sorry, I meant the code, not the output strings.

**Review by JustinChilton — APPROVED** [review]
